File: content/problems/2064-minimized-maximum-of-products-distributed-to-any-store/solutions/binary-search-store-capacity/Solution.kt
class Solution {
    fun minimizedMaximum(n:Int,quantities:IntArray):Int{
        var left=1;
        var right=quantities.maxOrNull()!!;
        while(left<right){
            val middle=(left+right)/2;
            var required=0L;
            for(quantity in quantities)required+=(quantity+middle-1)/middle;
            if(required<=n)right=middle else left=middle+1
        };
        return left
    }
}

File: content/problems/2064-minimized-maximum-of-products-distributed-to-any-store/solutions/binary-search-store-capacity/solution.cpp
class Solution {
    public: int minimizedMaximum(int n,vector<int>& quantities){
        int left=1,right=*max_element(quantities.begin(),quantities.end());
        while(left<right){
            int middle=(left+right)/2;
            long long required=0;
            for(int quantity:quantities)required+=(quantity+middle-1)/middle;
            if(required<=n)right=middle;
            else left=middle+1;
        }
        return left;
    }
};
